File: Study Plan/LeetCode 75/0394 Decode String/DecodeString.java
/*  
    Problem
    --------------------------------------------------------------------------------
    394. Decode String
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 1 ms (Beats 77.47% of users with Java)
    - Memory 41.12 MB (Beats 89.97% of users with Java)
*/

public class Solution {
    public String decodeString(String s) {
        String encoded = s;
        StringBuilder decoded = new StringBuilder();
        Stack<String> letters = new Stack<>();
        Stack<Integer> numbers = new Stack<>();
        int k = 0;
        for (char character : encoded.toCharArray()) {
            if (Character.isDigit(character)) {
                k = k * 10 + (character - '0');
            } else if (character == '[') {
                letters.push(decoded.toString());
                numbers.push(k);
                decoded = new StringBuilder();
                k = 0;
            } else if (character == ']') {
                String letter = letters.pop();
                int number = numbers.pop();
                StringBuilder data = new StringBuilder();
                for (int index = 0; index < number; index++) {
                    data.append(decoded);
                }
                decoded = new StringBuilder(letter).append(data);
            } else {
                decoded.append(character);
            }
        }
        return decoded.toString();
    }
}

File: Study Plan/LeetCode 75/0394 Decode String/DecodeString.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    394. Decode String
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 47 ms (Beats 91.46% of users with C#)
    - Memory 38.53 MB (Beats 64.90% of users with C#)
*/


public class Solution
{
    public string DecodeString(string s)
    {
        var encoded = s;
        var decoded = string.Empty;
        var letters = new Stack<string>(s.Length);
        var numbers = new Stack<int>(s.Length);
        var k = 0;

        foreach (char character in encoded)
        {
            if (char.IsDigit(character))
            {
                k = k * 10 + (character - '0'); 
            }
            else if (character == '[')
            {
                letters.Push(decoded);
                numbers.Push(k);
                decoded = string.Empty;
                k = 0;
            }
            else if(character == ']')
            {                  
                decoded = letters.Pop() + string.Join("", Enumerable.Repeat(decoded, numbers.Pop()));
            } 
            else
            {
                decoded += character;
            }
                
        }

        return decoded;
    }
}
